Update openshift origin client used in e2e tests to v3.11.0

amitkrout opened the pull request:
## What is the purpose of this change? What does it change?
<!-- Describe the changes here, as detailed as needed. -->
Test should be run on latest available OpenShift release version

## Was the change discussed in an issue?
Mattermost channel - https://chat.openshift.io/developers/pl/cn5ihkt8htdy5emjdeg89ribze
<!-- Please do Link issues here. -->

## How to test changes?
This is CI specific. On master test suite runs on oc version v3.10.0, however with this changes test suite will run on oc v3.11.0

amitkrout commented:
CI fails because the default output of v3.11.0 ```oc version``` does not contain ```openshift``` sub string which is part of odo version validation in the test suite. 
```
Running command: odo version
odo v0.0.17 (a3933167)
Server: https://127.0.0.1:8443
Kubernetes: v1.11.0+d4cacc0
Running command: oc version|grep Server|cut -d ' ' -f 2
https://127.0.0.1:8443
Running command: oc version|grep openshift|cut -d ' ' -f 2
Running command: oc version|grep kubernetes|tail -1|cut -d ' ' -f 2
v1.11.0+d4cacc0
• Failure [0.840 seconds]
odoe2e validate odo version cmd with oc version [It] should show the version of OpenShift server 
/home/travis/gopath/src/github.com/redhat-developer/odo/tests/e2e/e2e_test.go:575
  Expected
      <string>: odo v0.0.17 (a3933167)
      
      Server: https://127.0.0.1:8443
      Kubernetes: v1.11.0+d4cacc0
      
  to contain substring
      <string>: OpenShift
  /home/travis/gopath/src/github.com/redhat-developer/odo/tests/e2e/e2e_test.go:584
```

amitkrout commented:
> CI fails because the default output of v3.11.0 oc version does not contain openshift sub string which is part of odo version validation in the test suite.

CI failure is fixed. Now pr is open for review

amitkrout commented:
Did some experiment with test pattern but did not work out due to restriction in ```GinkGo``` defined interface. However finally landed in good test shape with implementing ```re``` package for validating version strings.

Unfortunately version check returns ```false``` causing the failure in CI. Will check tomorrow and hence marking it as WIP.

amitkrout commented:
Thanks a lot @syamgk for pointing some cool test debugging stuff on GinkGo test framework. 

> Unfortunately version check returns false causing the failure in CI. Will check tomorrow and hence marking it as WIP.

Issue is fixed. Removing WIP tag and opening it for review

anmolbabu commented:
Now that we intend to test odo against multiple openshift versions, can we as part of this PR parametrise the version and may be pass the versions from our travis job config directly or may be even via a intermediary script?(I am not sure about the practicality of it but may be worth a exploration?)

cdrage commented on scripts/oc-cluster.sh:
```diff
@@ -12,15 +12,15 @@ sudo service docker start
 
 ## download oc binaries
-sudo wget https://github.com/openshift/origin/releases/download/v3.10.0/openshift-origin-client-tools-v3.10.0-dd10d17-linux-64bit.tar.gz -O /tmp/openshift-origin-client-tools.tar.gz 2> /dev/null > /dev/null
+sudo wget https://github.com/openshift/origin/releases/download/v3.11.0/openshift-origin-client-tools-v3.11.0-0cbc58b-linux-64bit.tar.gz -O /tmp/openshift-origin-client-tools.tar.gz 2> /dev/null > /dev/null
```
Like what @anmolbabu said, perhaps we can make this link into a constant / variable? Take the link and put it in a $OPENSHIFT_URL_DOWNLOAD or something.

Same with the bellow code too. If we do that, we can possibly change the version we want to test against in the future..

surajnarwade replied:
+1 :+1:

geoand replied:
:+1:

kadel commented:
> Mattermost channel - https://chat.openshift.io/developers/pl/cn5ihkt8htdy5emjdeg89ribze

stuff like this should be discussed here in the issues :-(

amitkrout commented:
@cdrage @surajnarwade @geoand Hope i addressed the comment https://github.com/redhat-developer/odo/pull/1166#pullrequestreview-188512496. Please have a look 

I have verified in travis-ci
```
[…]
$ OPENSHIFT_CLIENT_BINARY_URL=https://github.com/openshift/origin/releases/download/v3.10.0/openshift-origin-client-tools-v3.10.0-dd10d17-linux-64bit.tar.gz ./scripts/oc-cluster.sh
docker stop/waiting
# we keep "-H tcp://..." for API clients that can't to talk to UNIX sock --insecure-registry 172.30.0.0/16ets
DOCKER_OPTS="-H tcp://127.0.0.1:2375 -H unix:///var/run/docker.sock --insecure-registry 172.30.0.0/16"
docker start/running, process 4386
docker start/running, process 4386
tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.security.selinux'
openshift-origin-client-tools-v3.10.0-dd10d17-linux-64bit/oc
tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.security.selinux'
openshift-origin-client-tools-v3.10.0-dd10d17-linux-64bit/README.md
tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.security.selinux'
openshift-origin-client-tools-v3.10.0-dd10d17-linux-64bit/LICENSE
oc v3.10.0+dd10d17
kubernetes v1.10.0+b81c8f8
features: Basic-Auth GSSAPI Kerberos SPNEGO
Unable to find image 'openshift/origin:v3.10.0' locally
v3.10.0: Pulling from openshift/origin
[…]
```
```

[…]
$ ./scripts/oc-cluster.sh
docker stop/waiting
# we keep "-H tcp://..." for API clients that can't to talk to UNIX sock --insecure-registry 172.30.0.0/16ets
DOCKER_OPTS="-H tcp://127.0.0.1:2375 -H unix:///var/run/docker.sock --insecure-registry 172.30.0.0/16"
docker start/running, process 4445
docker start/running, process 4445
tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.security.selinux'
openshift-origin-client-tools-v3.11.0-0cbc58b-linux-64bit/oc
tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.security.selinux'
openshift-origin-client-tools-v3.11.0-0cbc58b-linux-64bit/kubectl
tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.security.selinux'
openshift-origin-client-tools-v3.11.0-0cbc58b-linux-64bit/README.md
tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.security.selinux'
openshift-origin-client-tools-v3.11.0-0cbc58b-linux-64bit/LICENSE
oc v3.11.0+0cbc58b
kubernetes v1.11.0+d4cacc0
features: Basic-Auth GSSAPI Kerberos SPNEGO
Unable to find image 'openshift/origin:v3.11.0' locally
v3.11.0: Pulling from openshift/origin
[…]
```

geoand commented on docs/ci-reference.md:
````diff
@@ -0,0 +1,31 @@
+        - make test-main-e2e
+```
+
+Incase there is a need of running odo integration test on top of specific oc version, use env variable `OPENSHIFT_CLIENT_BINARY_URL` to pass the [released](https://github.com/openshift/origin/releases) oc client URL in `.travis.yaml`. For oc v3.10.0, use the configuration 
````
`Incase there is a need of running odo integration test on top of specific oc version` -> `If the need presents itself to run odo integration tests against a specific version of Openshift`

amitkrout replied:
Done

geoand commented on scripts/oc-cluster.sh:
```diff
@@ -3,6 +3,9 @@
 ## Inspired by https://github.com/radanalyticsio/oshinko-cli/blob/master/.travis.yml
 
+## Use this variable to get more control over downloading client binary
+OPENSHIFT_CLIENT_BINARY_URL=${OPENSHIFT_CLIENT_BINARY_URL:-'https://github.com/openshift/origin/releases/download/v3.11.0/openshift-origin-client-tools-v3.11.0-0cbc58b-linux-64bit.tar.gz'}
```
:+1:

girishramnani commented on scripts/oc-cluster.sh:
```diff
@@ -12,15 +15,15 @@ sudo service docker start
 
 ## below cmd is important to get oc working in ubuntu
-sudo docker run -v /:/rootfs -ti --rm --entrypoint=/bin/bash --privileged openshift/origin:v3.10.0 -c "mv /rootfs/bin/findmnt /rootfs/bin/findmnt.backup"
+sudo docker run -v /:/rootfs -ti --rm --entrypoint=/bin/bash --privileged openshift/origin:`echo $OPENSHIFT_CLIENT_BINARY_URL | awk -F '//' '{print $2}' | cut -d '/' -f 6` -c "mv /rootfs/bin/findmnt /rootfs/bin/findmnt.backup"
```
might wanna move this into a variable `OPENSHIFT_CLIENT_LATEST_VERSION`

amitkrout replied:
Good choice!!!
Done

surajnarwade commented:
waiting for @cdrage 's :red_circle:   to :green_heart:

cdrage commented on docs/ci-reference.md:
```diff
@@ -0,0 +1,31 @@
+# How to run integration test job in CI
```
How to run the integration test job in Travis CI

cdrage commented on docs/ci-reference.md:
```diff
@@ -0,0 +1,31 @@
+# How to run integration test job in CI
+
+For default oc, use the configuration in .travis.yaml. For example 
```
For example:

cdrage commented on docs/ci-reference.md:
````diff
@@ -0,0 +1,31 @@
+        - make test-main-e2e
+```
+
+If the need presents itself to run odo integration tests against a specific version of Openshift, use env variable `OPENSHIFT_CLIENT_BINARY_URL` to pass the [released](https://github.com/openshift/origin/releases) oc client URL in `.travis.yaml`. For oc v3.10.0, use the configuration
````
Odo not odo

cdrage replied:
use the configuration:

amitkrout commented:
Fixed @cdrage's review comment. 
@surajnarwade Can you please merge the pr
